app/auth/callback/page.tsx

```diff
@@ -0,0 +1,137 @@
+"use client";
+
+// This must be a client component, because Supabase sets the session from a fragment URL (#access_token=...) which is only accessible in the browser
+//   const router = useRouter();
+// if we want to redirect the user
+import { useEffect, useState } from "react";
+import { useRouter } from "next/navigation";
+import supabaseClient from "@/lib/supabase";
+
+// needs to be a page.tsx not a route.ts because supabaseClient.auth.getSession() is using client side logic
+// otherwise they'll be an error because route.ts is just for api handlers
+
+// https://supabase.com/docs/guides/auth/social-login/auth-google?queryGroups=environment&environment=client#google-pre-built
+
+export default function AuthCallback() {
+  const [messageToUser, setMessageToUser] = useState(
+    "Finishing signing you in ...",
+  );
+  const router = useRouter();
+
+  useEffect(() => {
+    // edge case if someone visits the raw http://localhost:3000/auth/callback page
+    // this prevents them from getting stuck forever on "Finishing signing you in..."
+    // since Supabase never receives valid tokens to process.
+
+    const urlParams = new URLSearchParams(window.location.search);
+    const hasAuthCode =
+      urlParams.has("code") || urlParams.has("error_description");
+
+    const isAuthCallbackEmpty =
```

Comment:
Last change (promise) @thomasnguyen 

finally replicated that issue luis mentioned where it gets stuck on the "finishing signing you in ..." message 

it happened if they visited with an invalid url http://localhost:3000/auth/callback
without "?code=62c29060-3a18-4daf-a1a6-7b44449bbdee" added a check where if this happens, they still get redirected

Before it was getting stuck because supabase only understands urls with ? and either "code" or "error_description"

Tested on both my devices with valid and invalid urls, its working